torch/autograd/gradcheck.py

```diff
@@ -64,22 +66,57 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
     for x_tensor, d_tensor in zip(x_tensors, j_tensors):
         # need data here to get around the version check because without .data,
         # the following code updates version but doesn't change content
-        x_tensor = x_tensor.data
-        for d_idx, x_idx in enumerate(product(*[range(m) for m in x_tensor.size()])):
-            orig = x_tensor[x_idx].item()
-            x_tensor[x_idx] = orig - eps
-            outa = fn(input).clone()
-            x_tensor[x_idx] = orig + eps
-            outb = fn(input).clone()
-            x_tensor[x_idx] = orig
-
-            r = (outb - outa) / (2 * eps)
-            d_tensor[d_idx] = r.detach().reshape(-1)
+        if x_tensor.is_sparse:
+            def get_stride(size):
+                dim = len(size)
+                tmp = 1
+                stride = [0] * dim
+                for i in reversed(range(dim)):
+                    stride[i] = tmp
+                    tmp *= size[i]
+                return stride
+
+            x_nnz = x_tensor._nnz()
+            x_size = list(x_tensor.size())
+            x_indices = x_tensor._indices().t()
+            x_values = x_tensor._values().data
+            x_stride = get_stride(x_size)
+
+            for i in range(x_nnz):
```

Comment:
I am not sure if this should be called `gradcheck` because you are only checking nonzero values rather than the full tensor. :/ I know that we only calculate gradients for the present values, but technically these are not the gradients you would expect for the normal definition of those ops. So having the gradcheck default to assuming this seems wrong to me. Could you add a flag to gradcheck that enables this and clarify in the doc?

Reply:
I see, this is a valid point. Can I make it `sparse_gradcheck` and add doc for it? Although most of the code in this fn will be dups of `gradcheck`. What do you think? I guess you also know why I am only checking nnz values here: I don't know how to permute zero values in a sparse tensor...

Reply:
I was thinking about adding a bool flag to `gradcheck` that would turn on/off checking only nonzero values for sparse inputs. I don't know if that's the best solution though.

Reply:
yes, that's doable too. I can add a `is_sparse` flag to `gradcheck`